### PR TITLE
add segment implementation of metametrics

### PR DESCRIPTION
**Pull request by brad-decker**
This borrows heavily from the previous PR: #9354, which was mainly implemented by @danjm. 

**Why a new PR?**
The code had changed significantly as well as the file structure, pushing to the previous branch would have been disingenuous and would have resulted in confusing diff with the feedback there. I did, however, consult the feedback on that PR while forming this one.

**What does it do?**
1. Adds a new `metametrics.new.js` file in the `contexts/` folder that started as a copy and paste of the original.
2. Through this new file and its associated `Context`, adds support for tracking events through segment, which is an analytics provider that connects to many different tools. Namely, mixpanel.
3. Adds a `SEGMENT_WRITE_KEY` environment variable that falls back to our dev environment write key. This hardcoded key likely needs to be hidden inside a .env file to avoid forks from sending events to our dev environment.
4. Adds page tracking functionality to the new `MetaMetricsProvider` using `matchRoute` from `react-router`. This allows us to send a known string as the name of the page.
5. Changes the API for sending an event using the `MetaMetrics` context **significantly**.

**New API for sending events**

Previous:
```js
this.context.metricsEvent({
  eventOpts: { category: 'X', action: 'X', name: 'X', anyOtherAttributes: 'x' },
  config: { excludeMetaMetricsId: true, isOptIn: true }
});
```

New:
```js
this.context.trackEvent({
  event: 'Unlocked MetaMask',
  category: 'X',
  excludeMetaMetricsId: true,
  isOptIn: true,
  properties: {
   some_snake_case_property: 'value'
  },
  include: ['number_of_tokens']
})
```

key differences:
1. the config object in the previous version has been lifted and flattened into the primary object
2. action prop is likely going away pending analytics tracking plan.
3. the required eventOpts event and category are lifted and flattened into the primary object
4. a specific properties object may be included for tracking additional properties
5. values that are available to the Context Provider, such as version, number_of_tokens can be included in the event by listing them as strings. This API allows for downstream consumers to access this data without having to tap into it everywhere. previously we were always sending all of those values with every event...now we can control which ones to send. Defaults to sending *none*.

### CODE REVIEW

**Comment by brad-decker** [comment]
@Gudahtt the lodash update in analytics-node wasn't a vulnerability update, but an update to axios was which is why the beta version is back on this branch.

**Comment by brad-decker** [comment]
Going to wait for #9368 to land so I can piggyback on the `.metamaskrc`for the development build segment write key. @whymarrh should I just include a spoofed version in the distributed rc file?

**Comment by whymarrh** [comment]
> should I just include a spoofed version in the distributed rc file?

Yeah an example value that maybe can even be used without changing it (for dev purposes?)

**Comment by brad-decker** [comment]
@MetaMask/extension-devs -- updated this to use .metamaskrc, in the dist file I included the variable with no value so that it'll be undefined unless explicitly provided. I tested the test build and it works as expected. There's a noop of the segment module that is exported if the environment variable is undefined.

**Review by whymarrh — COMMENTED** [review]
This LGTM, but I'll leave it to someone else (@danjm or @Gudahtt) to confirm the Segment usage

File: ui/app/contexts/metametrics.new.js
Comment (whymarrh) [inline ui/app/contexts/metametrics.new.js:112]: I think this works for properties too, to mark them as optional:

```suggestion
   * @param {boolean}  [config.isOptIn] - happened during opt in/out workflow
```

File: ui/app/contexts/metametrics.new.js
Comment (whymarrh) [inline ui/app/contexts/metametrics.new.js:113]:
```suggestion
   * @param {object}   [config.properties] - object of custom values to track, snake_case
```

File: ui/app/contexts/metametrics.new.js
Comment (whymarrh) [inline ui/app/contexts/metametrics.new.js:114]:
```suggestion
   * @param {string[]} [config.include] - array of values available in this provider that can be included
```

File: ui/app/contexts/metametrics.new.js
Comment (whymarrh) [inline ui/app/contexts/metametrics.new.js:137]: Nit:

```suggestion
```

File: ui/app/hooks/useMetricEvent.js
Comment (whymarrh) [inline ui/app/hooks/useMetricEvent.js:21]:
```suggestion
   * @param {object}   config - configuration object for the event to track
   * @param {string}   config.event - event name to track
   * @param {string}   config.category - category to associate event to
   * @param {boolean}  [config.isOptIn] - happened during opt in/out workflow
   * @param {object}   [config.properties] - object of custom values to track, snake_case
   * @param {string[]} [config.include] - array of values available in this provider that can be included
```

(I remembered I can do multiline suggestions 😅)

**Review by whymarrh — COMMENTED** [review]
I left a few small comments, overall this still LGTM!

> New API for sending events

This API makes sense to me!

**Comment by brad-decker** [comment]
thanks @whymarrh for the suggestions! Added them all.

File: development/build/scripts.js
Comment (Gudahtt) [inline development/build/scripts.js:337]: We should probably use a separate key for production builds v.s. non-production (i.e. PR or release candidate) builds.

Maybe one for prod, one for everything else? :thinking: Or three, with one one additional key for RCs or any test builds that we distribute.

Or is there some other way to separate test metrics from production? Sentry has the `environment` property for example.

Comment (brad-decker) [reply development/build/scripts.js:337]: There's a development key that we could use for PRs, which I think would be fine. The purpose of the development source in Segment is to debug events, and ensure that they are tracked to plan. I'm curious if we could override SEGMENT_WRITE_KEY in CI for release candidate builds?

Would the RC builds have a special version number returned from ```platform.getVersion()``` or no?

Comment (Gudahtt) [reply development/build/scripts.js:337]: I don't know of a way to change the CI environment variable in different scenarios. There's just one set of variables for the whole project. Using two separate keys seems easiest. We can still collapse them to `SEGMENT_WRITE_KEY` when we inject it with `envify`.

The only way to identify the RC builds would be the `environment` variable (injected as `METAMASK_ENVIRONMENT`). That should say whether it's a release candidate, a PR, a "staging" build (i.e. `develop`), or something else (e.g. a local build made with `yarn dist`).

Comment (Gudahtt) [reply development/build/scripts.js:337]: Just using the dev key for anything non-production seems OK to me for now though! Adding the third key was just an idea. We can revisit it if we think that would be useful.

Comment (brad-decker) [reply development/build/scripts.js:337]: handled in https://github.com/MetaMask/metamask-extension/pull/9382/commits/af67e9661b14ae050cda6a65e8d8d845225bc03f

Comment (brad-decker) [reply development/build/scripts.js:337]: This now includes network, platform, environment, environment_type and version in context instead of properties.

Comment (brad-decker) [reply development/build/scripts.js:337]: well shit, I guess this doesn't actually solve the problem of where the events land -- I do think PRs should hit development. RCs probably should go to prod? I'll ask Lorenzo.

Comment (Gudahtt) [reply development/build/scripts.js:337]: I would suggest that prod only be "real users", on official builds. Anyone using a different build is likely a developer or a tester. Until we create a separate dashboard for RCs, I think dev would make sense. The dev dashboard is likely to be low-volume enough to find relevant metrics from the RCs anyway.

Comment (brad-decker) [reply development/build/scripts.js:337]: that works for me

Comment (Gudahtt) [reply development/build/scripts.js:337]: The `context` solution looks great! That certainly simplifies things.

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [inline ui/app/helpers/constants/routes.js:67]: Whoops - this route is unused. I can remove it in a separate PR.

File: development/build/scripts.js
Comment (brad-decker) [reply development/build/scripts.js]: https://github.com/MetaMask/metamask-extension/pull/9382/commits/e912c0b0c9892219a130d5b3681c5e48fb4799a3 final piece of the puzzle, i think

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [inline ui/app/helpers/constants/routes.js:74]: Is this variable name (`address`) used anywhere? :thinking: I noticed that we use `:id` for this route elsewhere, but I don't think it matters, and `address` is more descriptive.

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [inline ui/app/helpers/constants/routes.js:94]: I think we should have `${CONFIRM_TRANSACTION_ROUTE}/:id` here as a route as well? :thinking: 

It would immediately redirect to one of these other routes, but it's still on that page until the redirect happens, during the render.

Comment (brad-decker) [reply ui/app/helpers/constants/routes.js:94]: I can add it, but we'd want to skip that page track unless it has some importance to the user flow? I have handled this route specifically in the page tracking code to skip tracking the root page before the redirect.

File: ui/app/helpers/constants/routes.js
Comment (brad-decker) [reply ui/app/helpers/constants/routes.js:74]: oh hmm, this actually is probably worth calling out and maybe preventing -- It is used, in the page track call we pass the params from the matchRoute function which when a user visits `${CONTACT_EDIT_ROUTE}/${my_address}` would send this to segment:

```
params: { address: my_address }
```

this would allow us to identify accounts with metametrics id. This seems like we definitely want to avoid this. Maybe we don't need to send the route params to segment at all? it seemed cool for the asset page where the params would be the token address or ETH

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [reply ui/app/helpers/constants/routes.js:94]: Oh I see, I only just noticed this. This seems fine, nevermind!

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:52]: they were being attached to all events for Matomo, but i doubt they will be in the tracking plan to be on every event. Do any of these stand out as things that make more sense as context instead of properties? I will remove the rest and put that responsibility on the downstream event calls. Didn't really like this API but didn't want to force more selector/hook usage elsewhere.

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [reply ui/app/helpers/constants/routes.js:74]: Ah, good point, yes we should definitely not send addresses. I agree that the asset seems useful though :thinking: 

I wonder is there a way to send one but not the other?

Comment (brad-decker) [reply ui/app/helpers/constants/routes.js:74]: i'm going to change 'account' to 'address' for the my_account routes, and then omit that from params.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:52]: Good question - I'm not sure.

`network` is fairly important to most metrics, as it lets us hone in on "real" transactions and exclude local and testnet stuff. So maybe that makes sense in context. I'm a little torn only because we intend to make network a per-dapp setting, but even still there might be a "current network" for any given UI state? Or at least most states. :thinking: We can cross that bridge when we come to it though. We can add a `N/A` value or something later if necessary.

`activeCurrency` is useful to know in general, but I don't know that we need it per-event? That seems like something that could go in `identify`. The same goes for `number_of_accounts` and `number_of_tokens`. `accountType` was meant to track how common hardware wallets were, and we could solve that with `identify` as well (e.g. `number_of_hardware_accounts` perhaps?).

`dapp_url` is critical for confirmations, but non-applicable anywhere else. I guess that could go in `context` too, and it'd just be blank for the non-applicable cases? I don't feel strongly as long as we track it somehow.

File: ui/app/helpers/constants/routes.js
Comment (brad-decker) [reply ui/app/helpers/constants/routes.js:74]: https://github.com/MetaMask/metamask-extension/pull/9382/commits/f20ab97e0739bdad960bdf81c4a322ff9f9e80da i ended up not changing the param name, but instead omitted both

File: ui/app/helpers/constants/routes.js
Comment (brad-decker) [reply ui/app/helpers/constants/routes.js:67]: ill remove it from the PATH_NAME_MAP in this one.

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:52]: `dapp_url` does show up in the mobile stuff as event properties, so probably not context. We could add it to the confirmation screens and include them in the events manually. That probably makes more sense.

When does `activeCurrency` change? 

the other variables I agree on identify calls as a means to track.

Already added network to context so it definitely doesn't need to be there. Cool.

File: ui/app/helpers/constants/routes.js
Comment (brad-decker) [reply ui/app/helpers/constants/routes.js]: done

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:52]: Oops, I was mixing up `activeCurrency` with `currentCurrency`.

`currentCurrency` is the user's preferred currency for display, and might be useful in `identify` along with the current locale.

`activeCurrency` appears to be the current selected currency when in the send flow specifically, which doesn't seem appropriate for context or identify. That seems like it should be explicitly passed in by the send flow events.

Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: resolved with https://github.com/MetaMask/metamask-extension/pull/9382/commits/34c92162e19d72a15eaaff438f4b94aae02ee9c7

Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: and i'll fix the lint issues I overlooked in the next commit ;)

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:65]: The `url` here would also contain the same private information being excluded below (the user's address)

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:160]: Yes, I also think this is true :sweat_smile:

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:88]: Nit: leftover `console.log`

We should consider adding this to the lint rules :thinking:

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: changed to use 'path' which will have `:address` instead of `0x0000000000000000` -- for those URLs with params objects without sensitive data in them we already get the valuable pieces.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:120]: I'm not sure this will work :thinking: 

https://segment.com/docs/connections/spec/common/#context

>Context is a complete and explicit specification, so properties outside the spec are ignored. You should only use Context fields for their intended meaning.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:55]: Could you add the same `context` properties to the page calls as well? Or do you think there's a good reason to exclude them?

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:120]: Perhaps these could just always be added to `properties`? At least for those without an equivalent `context` property.

Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:120]: hmm i'll test it out and see if they end up in segment.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:116]: Nit: We should consider disallowing the following property names from being passed in via this "properties" parameter: `revenue`, `currency`, and `value`. These are [reserved property names](https://segment.com/docs/connections/spec/track/#properties) and are assumed to have a specific meaning by Segment. `value`especially seems easy to accidentally use for some other purpose.

Passing these reserved names in via distinct config properties seems safer? At least then we can add a JSDoc description explaining what to use it for.

File: ui/app/helpers/constants/routes.js
Comment (Gudahtt) [reply ui/app/helpers/constants/routes.js]: Removed here: https://github.com/MetaMask/metamask-extension/pull/9394

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:120]: image they come through segment but then don't make it into mixpanel.

Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:120]: @Gudahtt -- I think I can put version in app, so 

```
app: {
  version: platform.getVersion(),
  build: process.env.METAMASK_ENVIRONMENT,
},
```

network maybe goes into an identify call or attached as a property to every event? environment_type won't be able to go anywhere. I added userAgent which correctly identifies the browser in mixpanel so we can avoid that one.

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: I did exactly this. Check the JSDoc to make sure my language is ok.

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: removed.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:122]: Putting those two things in `app` sounds good! Though for `build` it'd be nice to use the git hash instead :thinking: I had been meaning to add that to Sentry as well, and the git hash would make for a better build number. Many verison + environment combinations would have multiple builds.

Maybe it would be more appropriate to put the environment in the version? e.g. something like

```
let version = platform.getVersion()
if (process.env.METAMASK_ENVIRONMENT !== 'production') {
  version = `${version}-${process.env.METAMASK_ENVIRONMENT}`
}
```

Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:122]: `network` doesn't seem like a great candidate for identify, since it can change from minute-to-minute. I would go the property route. `environment_type` seems like it could be a property as well :thinking:

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js]: Looks good to me!

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:85]: added in https://github.com/MetaMask/metamask-extension/pull/9382/commits/ec698f97a626495b999bbba0641881e0ba2bb4e0 -- I moved the context building logic into a separate hook within the same file to keep it slightly more clean. Let me know what you think.

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: done in https://github.com/MetaMask/metamask-extension/pull/9382/commits/f5f1c3d7d24e1cd4fcd68b1ee0dbe06017a949d9

**Comment by brad-decker** [comment]
@Gudahtt i think your feedback is all addressed, i made a mess of the commits though. the last one adds some failsafes for not tracking duplicate page calls which I noticed on popups and notifications. It would track the home route first, then the confirmation / send-ether route (or whichever) three times for some reason.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:168]: Is this useful to track? I don't think we tracked it before - we just used this to determine whether to anonymize the event

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:63]: Hey this is a really interesting idea - using the dapp that triggered the confirmation as the referrer. Not _exactly_ what this field is usually used for, but it feels right semantically :thinking:. I think I like this.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:61]: We use underscores instead of dashes for the locale code, which is not compliant with the language tag spec. It looks like Segment also expects dashes instead of underscores ([see the `locale` field in the context docs](https://segment.com/docs/connections/spec/common/#context)). So we should do a substitution here before we send this in.

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:87]: This will still send "anonymous" calls in during initialization even after explicit opt-out. This is not a _huge_ deal, but we could prevent this for checking whether `participateInMetaMetrics` is `false` or completely unset (the default value is `null`).

e.g.:

```suggestion
    if (
      (participateInMetaMetrics === null && location.pathname.startsWith('/initialize'))
      || participateInMetaMetrics
    ) {
```

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [reply ui/app/contexts/metametrics.new.js:100]: Looks good to me!

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:96]: We do want _subsequent_ navigations back to home to track though :thinking: Or at least I'd think we would. This seems to skip tracking _any_ navigations to home.

Would this still serve the purpose you had in mind?

```suggestion
        // If we're in a popup or notification we don't want the initial home route to track
        !(
          (environmentType === 'popup' || environmentType === 'notification')
          && match.path === '/'
          && previousMatch.current === undefined
        )
```

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:165]: It's now impossible to send these values even via `config` - nothing seems to be done with them, unless I'm missing something.

We'd want these three values added to `properties` here if defined, right?

File: ui/app/contexts/metametrics.new.js
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:108]: This property seems to be missing from the `properties` of the `track` call

File: ui/app/contexts/metametrics.new.js
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js:175]: In retrospect, no -- you get the same information by seeing if an anonymousId or userId was provided.

**Comment by brad-decker** [comment]
@Gudahtt your most recent feedback is now addressed.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**
